---
name: Achieved standards
about: Monitor achieved assembly standards
title: "[Standards] Species Name"
labels: ''
assignees: ''

---

## Standards achieved

- [ ] **contig NG50 > 1 MB**: Ref standard > 6.C.Q40 ( X.Y.Q = contig NG50 log10 value; scaffold NG50 log10 value; QV value; C denotes telomere-to-telomere)
- [ ] **scaffold NG50 = telomere-to-telomere **: Ref standard > 6.C.Q40 ( X.Y.Q = contig NG50 log10 value; scaffold NG50 log10 value; QV value; C denotes telomere-to-telomere)
- [ ] **Merqury QV > 40**: Ref standard > 6.C.Q40 ( X.Y.Q = contig NG50 log10 value; scaffold NG50 log10 value; QV value; C denotes telomere-to-telomere)
- [ ] **Busco: > 90% complete genes and single copy**
- [ ] **Merqury: < 5% false duplications**
- [ ] **Merqury: > 90% kmer completeness**
- [ ] **Samtools: > 90% sequences assigned to candidate chromosomal sequences**
- [ ] **Star: > 90% transcripts from the same organism mappable**
- [ ] **Separation of sequence of the target species from contaminants and other organisms such as symbionts/cobionts**
- [ ] **Explicit identification of a primary (haploid or pseudo-haploid) assembly, with additional sequence in a secondary bin that may contain either full alternate haplotypes or a set of haplotypic/other sequence from the individual**
- [ ] **Separation and explicit identification of organellar genomes**
- [ ] **Only A,C,G,T and N bases and sequences should not begin or end with Ns**
- [ ] **Identification of discordances between raw data and resulting assembly to locate and remove structural errors (misjoins, missed joins and false duplications)**
- [ ] **Identification and naming of chromosomes, esp. sex chromosomes, where possible**
- [ ] **Reconciliation with the known karyotype where it exists and this is possible**
